File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        self.max_count = 0
        self.curr_count = 0
        def dfs(x, y):
            if grid[x][y] == 1:
                self.curr_count += 1
                grid[x][y] = 0
                if x - 1 >= 0:
                    dfs(x-1, y)
                if y - 1 >= 0:
                    dfs(x, y-1)
                if x + 1 < len(grid):
                    dfs(x+ 1, y)
                if y + 1 < len(grid[0]):
                    dfs(x, y+1)
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                if grid[x][y] == 1:
                    self.curr_count = 0
                    dfs(x, y)
                    self.max_count = max(self.max_count, self.curr_count)
        return self.max_count





























        # def dfs(x, y):
        #     if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[x]):
        #         return 0
        #     else:
        #         if grid[x][y] == 1:
        #             grid[x][y] = 0
        #             return 1 + (dfs(x - 1, y) + dfs(x + 1, y) + dfs(x, y + 1) + dfs(x, y - 1))
        #         else:
        #             return 0
            
        # m = 0
        # for x in range(len(grid)):
        #     for y in range(len(grid[x])):
        #         if grid[x][y] == 1:
        #             m = max(m, dfs(x, y))

        # return m

File: Data Structures & Algorithms/count-number-of-islands/submission-2.java
class Solution {
    public void dfs(char[][] grid, int x, int y)
    {
        int[][] d = { {1, 0}, {-1, 0}, {0, 1}, {0,-1}};
        for(int i = 0; i < d.length; i++)
        {
            int dx = x + d[i][0];
            int dy = y + d[i][1];
            if(dx >= 0 && dx < grid.length && dy >= 0 && dy < grid[0].length && grid[dx][dy] == '1')
            {
                grid[dx][dy] = '0';
                dfs(grid, dx, dy);
            }
        }

    }
    public int numIslands(char[][] grid) {
        int count = 0;
        for(int x = 0; x < grid.length; x++)
        {
            for(int y = 0; y < grid[0].length; y++)
            {
                if(grid[x][y] == '1')
                {
                    count++;
                    grid[x][y] = '0';
                    dfs(grid, x, y);
                }
            }
        }
        return count;
    }
}
